style: Make the whole style crate independent of the implementation

emilio opened the pull request:
This allows, among other things, having different implementations for parsing pseudo{elements, classes} in both `ports/geckolib` and in servo.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9567)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/selector_matching.rs, components/style/selector_impl.rs, components/style/matching.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style, layout, and script code, but no tests are modified. Please consider adding a test!

ecoal95 commented on components/style/selector_impl.rs:
```diff
@@ -81,11 +79,12 @@ impl SelectorImpl for ServoSelectorImpl {
+        if name.chars().all(|c| c.is_lowercase()) {
+            Ok(PseudoElement::from(name))
+        } else {
+            Ok(PseudoElement::from(&*name.to_ascii_lowercase()))
```
@bholley: Please look at the commit message, do you thing we could do well accepting only names that start with a dash, and it that case assume they are already lowercase?

ecoal95 commented:
Also note that if we're just planning to support the gecko anon-box pseudo-element list, we could keep `PseudoElement` as an enum and use this list: https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h

cc: @SimonSapin

emilio committed:
Refactor style to be completely backend-independent

This commit refactors the style crate to be completely independent of
the actual implementation and pseudo-elements supported.

This also adds a gecko backend which introduces parsing for the
anonymous box pseudo-elements[1], although there's still no way of
querying them.

https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h

ecoal95 commented:
Went with the second approach, and saw why both of them needed extra work.

Actually I'll need to implement a way of getting matching rules from a `SelectorMap` that only correspond to a single pseudo-element (or making an extra map per pseudo, which seems a no-go).

I can:

 * Do a post-traversal cleanup after getting all matching rules
 * Implement a function for that in rust-selectors (something like `get_all_matching_rules_checked`)
 * Extend the current `get_all_matching_rules`.

I'd go with the second, but I'd prefer to check-in with Simon in order to confirm that the current approach is correct, and also to discuss a design (in case it is needed) and prevent wasting work.

That being said, it's late, he's probably sleeping and I should too, so... enough for today! :P

SimonSapin commented on components/style/selector_impl.rs:
```diff
@@ -8,6 +8,73 @@ use selectors::parser::{ParserContext, SelectorImpl};
     After,
+    FirstLine,
+
+    // https://mxr.mozilla.org/mozilla-central/source/layout/style/nsCSSAnonBoxList.h
```
Why is all thing going here rather than in a separate `SelectorImpl` impl in `ports/geckolib`? Wasn’t that the point of making the parser generic?

ecoal95 replied:
Yeah, forgot to mention that.

Would it be fine to have a feature-gated gecko selector implementation inside style?

I say it because I tried doing exactly that, and the code is pretty suboptimal (and way less readable).

For example, this:

```rust
let map = match pseudo_element {
    Some(PseudoElement::Before) => &self.before_map,
    Some(PseudoElement::After) => &self.after_map,
    Some(_) => &self.other_pseudo_elements_map,
    None => &self.element_map,
};
```

Becomes this:

```rust
let map = if let Some(pseudo) = pseudo_element {
    if pseudo == E::Impl::parse_pseudo_element(&ParserContext::new(), "before").unwrap() {
        &self.before_map
    } else if pseudo == E::Impl::parse_pseudo_element(&ParserContext::new(), "after").unwrap() {
        &self.after_map
    } else {
        &self.other_pseudo_elements_map
    }
} else {
    &self.element_map
};
```

SimonSapin replied:
> Would it be fine to have a feature-gated gecko selector implementation inside style?

I’m not opposed but I don’t see how it would help.

> Becomes this: [...]

Yes, calling parsing code here is silly. It shows a mis-match between `PseudoElement` being generic and the `before_map` and `after_map` fields of `Stylist` being hard-coded. I don’t think this means me should abandon the generic parser and have a single `PseudoElement` type. I’d rather change the way `Stylist` is organized. Maybe some map of `PseudoElement` to `PerPseudoElementSelectorMap`?

ecoal95 replied:
Yes, but having a `PerPseudoElementSelectorMap` per possible pseudo seems overkill to me, but it might be doable, and it's, in fact, more ellegan than what we have right now.

I can try to do it first and get some numbers if you want.

I was talking about having two different `PseudoElement` types, as with two different `SelectorImpl`, but making a default impl feature gated, something like:

```rust
enum ServoPseudoElement {
    Before,
    After,
    FirstLine,
}

enum GeckoPseudoElement {
    Before,
    After,
    FirstLine,

     // ...
}

struct ServoSelectorImpl;
struct GeckoSelectorImpl;

impl ... for GeckoSelectorImpl { ... }
impl ... for ServoSelectorImpl { ... }

#[cfg(feature = "geckolib")]
pub type DefaultSelectorImpl = GeckoSelectorImpl;
#[cfg(not(feature = "geckolib"))]
pub type DefaultSelectorImpl = ServoSelectorImpl;
```

And we'll ve moving again with concrete types, and would work as long as both implementations have a `Before` and `After` variants.

SimonSapin replied:
Re overkill, I’m also open to changing more significantly how `Stylist` / `PerPseudoElementSelectorMap` / `SelectorMap` are organized. (Though I don’t have a specific design in mind right now.)

Re feature-gated impls, ok, I see what you mean. I’m not necessarily opposed to this, but it doesn’t seem nice. I’d like to explore alternatives some more.

Ms2ger commented:
I'm not happy about parsing pseudo-elements that Gecko doesn't support.

Ms2ger commented:
Should be "... that Servo doesn't support".

SimonSapin commented:
@Ms2ger same here. This is what I mean by a separate `SelectorImpl` impl in inline comments.

ecoal95 commented:
Could you guys take a quick look at this to see if the current approach is good enough?

Right now I've got servo building and working fine (with a rust-selectors update to add the `Eq` and `Hash` constraints to the `PseudoElement` associated type).

Will like to make a few more tweaks before landing it is it's good though (fnv and trying to not pre-insert elements in the `ApplicableDeclarationsCache`).

No gecko backend yet, I'll add it when I've got the time :P

emilio committed:
style: Use fnv hashing for pseudo-elements

emilio committed:
geckolib: Fix :link and :any-link pseudo-element matching

The logic was switched between them.

ecoal95 commented:
I think this should be mostly ready for landing once https://github.com/servo/rust-selectors/issues/69 lands...

r? @SimonSapin (for the style side)
r? @bholley (for the geckolib side)

Still missing the hability to query on-demand for anon-box pseudo-elements, but I'd prefer to do that as a followup since this is already a pretty decent patch.

SimonSapin commented on components/style/selector_impl.rs:
```diff
@@ -82,10 +101,43 @@ impl SelectorImpl for ServoSelectorImpl {
+        let pseudo_element = match_ignore_ascii_case! { name,
+            "before" => Before,
+            "after" => After,
+            "first-line" => FirstLine,
```
Why add `first-line` here? We shouldn’t parse things we don’t support.

ecoal95 replied:
Sure, sorry! Just parsed it in order to ensure that the problem with the first approaches was solved.

SimonSapin commented on components/style/selector_matching.rs:
```diff
@@ -179,11 +171,14 @@ impl Stylist {
-                            Some(PseudoElement::After) => &mut after_map,
+                        let map = if let Some(ref pseudo) = selector.pseudo_element {
+                            self.pseudos_map.entry(pseudo.clone())
+                                            .or_insert(PerPseudoElementSelectorMap::new())
```
Is `PerPseudoElementSelectorMap::new()` expensive? Maybe use `.or_insert_with(PerPseudoElementSelectorMap::new)` instead.

ecoal95 replied:
I think it should only be called on the empty case, but `or_insert_with` is more elegant anyways so I'll change it :P

SimonSapin commented on components/style/selector_matching.rs:
```diff
@@ -256,19 +251,18 @@ impl Stylist {
                 "Style attributes do not apply to pseudo-elements");
 
         let map = match pseudo_element {
+            Some(ref pseudo) => self.pseudos_map.get(pseudo).unwrap(),
```
Can this `.unwrap()` really not fail?

ecoal95 replied:
That's why we pre-insert the eagerly cascaded pseudos map in the stylist constructor, so yes, this can't fail.

SimonSapin replied:
Can’t this method be called with a non-eagerly-cascaded pseudo-element?

SimonSapin commented on components/style/matching.rs:
```diff
@@ -526,20 +534,16 @@ pub trait ElementMatchMethods<'le> : TElement<'le> {
         applicable_declarations.normal_shareable &&
-        applicable_declarations.before.is_empty() &&
-        applicable_declarations.after.is_empty()
+        applicable_declarations.per_pseudo.values().all(|v| v.is_empty())
```
We have to be careful here: pseudo-elements that have UA styles with a `*|*` selector risk defeating style sharing completely.

What pseudo-elements are likely to be in `applicable_declarations.per_pseudo` at this point? Only the eagerly-cascaded ones?

ecoal95 replied:
Actually there's no support for querying those kind of pseudo-elements, so just the eagerly cascaded ones will be here, which are filled a few lines before.

ecoal95 commented:
Updated rust-selectors, adressed the review comments, and squashed everything in commits that compiled.

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 5dfe480 with merge 5ec3557...

SimonSapin commented:
`components/style` looks good to me.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/1618)

ecoal95 commented:
Forgot to update test-unit, should be solved now. WPT and CSS tests seem to be passing.

bholley commented:
Reviewed 11 of 20 files at r4, 11 of 17 files at r5, 15 of 17 files at r6.
Review status: 37 of 41 files reviewed at latest revision, 16 unresolved discussions.

---

*[components/layout/traversal.rs, line 32 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGv97D4_Fb4h1eiPFX:-KAGv97D4_Fb4h1eiPFY:85522578) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/traversal.rs#L32)):*
Hm, this doesn't seem right. Why do we need the servo-specific Impl here? Shouldn't we just set up the traits such that TElement implements Element<Impl=SelectorImplExt> and introduce ServoSelectorImpl in ServoLayoutElement?

 This doesn't matter in the near term, but I'd like to keep the code general (if possible) since we might hypothetically plug Servo Layout into Gecko down the line. If we _are_ going to make this servo-specific for now though, it seems like we should instead just stop parameterizing on N.

---

*[components/layout/wrapper.rs, line 669 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGyDT6OAvuCN7zMsI5:-KAGyDT7VJIZspivXkZ4:1712736981) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/wrapper.rs#L669)):*
This seems clunky (here and below). As discussed elsewhere, I think this would make more sense if the HashMap values weren't Optional.

---

*[components/style/data.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGxx9FNRaeuU1l8D2o:-KAGxx9FNRaeuU1l8D2p:1247514336) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/data.rs#L17)):*
Why is the hashmap Option-valued? Should we represent the lack of ComputedValues by simply not having an entry in the hashmap at all?

---

*[components/style/dom.rs, line 172 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGswbnsfBx5urN810C:-KAGswbnsfBx5urN810D:-1179112668) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/dom.rs#L172)):*
Why do you need the explicit lifetime here?

---

*[components/style/restyle_hints.rs, line 83 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH-8LhtqEw7qsT5dmk:-KAH-8LhtqEw7qsT5dml:1559776243) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/restyle_hints.rs#L83)):*
This is kind of annoying (here and below). Can't we have E: ElementExt, where ElementExt inherits Element with Impl=SelectorImplExt?

---

*[components/style/restyle_hints.rs, line 107 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH30qIb0PTHfrnA-Mv:-KAH30qIb0PTHfrnA-Mw:-1125836788) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/restyle_hints.rs#L107)):*
I can't quite remember, but do you really need the Self here?

---

*[components/style/selector_impl.rs, line 5 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGtryRRk4k2c5lSuI9:-KAGtryRRk4k2c5lSuIA:1482087332) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/selector_impl.rs#L5)):*
At some point it would be good to move this and some of the other servo-specific stuff out of components/style, so that we don't accidentally use it from Gecko. We can worry about this later though.

---

*[components/style/traversal.rs, line 46 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH4XPAJ265xZ5cx5qb:-KAH4XPAJ265xZ5cx5qc:1811479336) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/traversal.rs#L46)):*
Can we make TElement constrain its implementation of Element with Impl=SelectorImplExt? I think that would significantly reduce the added type complication in this file.

---

*[ports/geckolib/selector_impl.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAHNDEnn4SaAOQNuUDe:-KAHNDEoU05iQwmb7Wxe:2046906741) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/selector_impl.rs#L17)):*
Is the ::prefix necessary / desirable? Not sure what the preferred Rust style is, but I tend to avoid it in C++.

---

*[ports/geckolib/traversal.rs, line 68 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH6PrRFwwEuI4Noppf:-KAH6PrSRiOfrxPziiYi:84045880) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/traversal.rs#L68)):*
As in the Servo case, it seems like the Impl can be an associated type on the trait, and that we don't actually need to introduce Gecko-specificity here. Not a huge problem if we do, though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9567)*
<!-- Sent from Reviewable.io -->

emilio committed:
style: Prevent a crash when push_applicable_declarations is called with a non eagerly-cascaded pseudo-element

This actually didn't happen, but this should be done sooner than later.

emilio committed:
style: Refactor the per_pseudo map from StyleData to avoid having an option value type.

This make the layout code way clearer.

ecoal95 commented:
Review status: 32 of 43 files reviewed at latest revision, 15 unresolved discussions.

---

*[components/layout/traversal.rs, line 32 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGv97D4_Fb4h1eiPFX:-KAHfz9z99AhWpdjU7lb:-827731309) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/traversal.rs#L32)):*
Nope, it's needed (here and in geckolib), because the `SharedLayoutContext` contains a `SharedStyleContext<ServoSelectorImpl>`.

Making it generic will involve a lot more work, which I'd prefer to do in a follow-up, if it's not a big deal :)

---

*[components/layout/wrapper.rs, line 669 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGyDT6OAvuCN7zMsI5:-KAHk4yrFG-X64itR5Xr:1254487415) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/wrapper.rs#L669)):*
Hum... Yeah, remember to have thought about that.

The only thing why it's an Option laid in `cascade_node_pseudo_element` (components/style/matching.rs).

I still did not have servo building again, so I focused on getting it building and forgot about that.

Just pushed a patch to refactor how `cascade_node_pseudo_element` works and prevents the hashmap value type to be an `Option`.

---

*[components/style/data.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGxx9FNRaeuU1l8D2o:-KAHlReYYEkAt6WZwOw5:346116021) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/data.rs#L17)):*
Same thing.

---

*[components/style/dom.rs, line 172 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGswbnsfBx5urN810C:-KAI4G5m0j-dQUkd30PB:1818102497) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/dom.rs#L172)):*
For some reason the compiler doesn't seem to be able to infer it :/.

You can check it if you remove the lifetime.

---

*[components/style/restyle_hints.rs, line 83 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH-8LhtqEw7qsT5dmk:-KAI5Hv8L4IvtXzWt74E:-33224112) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/restyle_hints.rs#L83)):*
I did try that, and I wasn't able to figure out a good way to do this without changing too much code.

If I use something like:

```rust
pub trait ElementExt: Element 
    where <Self as Element>::Impl: SelectorImplExt {...}
```

I should specify the `SelectorImplExt` in a lot more places than now.

Is there something I'm missing?

---

*[components/style/restyle_hints.rs, line 107 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH30qIb0PTHfrnA-Mv:-KAI4g-nMmBnIHI8vCnw:1290146658) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/restyle_hints.rs#L107)):*
Yes, it is needed.

---

*[components/style/selector_impl.rs, line 5 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGtryRRk4k2c5lSuI9:-KAI9GbzAeGYYQMVKtXZ:-896829500) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/selector_impl.rs#L5)):*
Agreed. I wouldn't know where to put that right now, and it could be a perfect `E-Easy`, so I think it's better to leave it as a follow-up.

---

*[components/style/selector_matching.rs, line 264 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9567#-K9yPFO5zxLehGqOS1yZ-r5-264:-KAHwaaxQxwVngrBCrmq:1408459306) ([raw file](https://github.com/servo/servo/blob/545e884b98189de0722bbe23dfb82b4327f78611/components/style/selector_matching.rs#L264)):*
Well, it actually isn't, but I guess it could be (and in fact it should be how gecko should do it if I'm not wrong).

Updated to reflect that.

---

*[components/style/traversal.rs, line 46 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH4XPAJ265xZ5cx5qb:-KAHl_5RJH6OZGMoMzjH:-1023130172) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/traversal.rs#L46)):*
The same as above (`Element where Self::Impl: SelectorImplExt`) happens.

---

*[ports/geckolib/selector_impl.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAHNDEnn4SaAOQNuUDe:-KAHlgAnz1tn-ChmovKH:2090657438) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/selector_impl.rs#L17)):*
Well, I could `use style;` to remove the prefix, but I don't think it matters too much. Never read anything specifically against it.

---

*[ports/geckolib/traversal.rs, line 68 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH6PrRFwwEuI4Noppf:-KAI9dHxH3QgzXRhX3HN:1564324033) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/traversal.rs#L68)):*
Same reason as above, `SharedStyleContext` is really `style::context::SharedStyleContext<GeckoSelectorImpl>`, and as such both implementations must match.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9567)*
<!-- Sent from Reviewable.io -->

bholley commented:
Review status: 32 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/traversal.rs, line 32 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGv97D4_Fb4h1eiPFX:-KAIXMfViovTSchZEdCF:-1730757625) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/traversal.rs#L32)):*
Ok. If you have immediate plans to do a followup, feel free to leave as-is. Otherwise, I'd prefer to just operate on ServoLayoutNode and remove the generic stuff here, because it isn't actually generic given the Impl=ServoSelectorImpl.

---

*[components/style/restyle_hints.rs, line 83 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAH-8LhtqEw7qsT5dmk:-KAIc0v_TWo_HFJiG-ON:1713467416) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/style/restyle_hints.rs#L83)):*
Hm, I thought there was a way for traits to add bounds to the associated types of their bounds, but I can't seem to figure it out right now. Anyway, not important, so let's not worry about it for now.

---

*[ports/geckolib/selector_impl.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAHNDEnn4SaAOQNuUDe:-KAIaXgUsWdfSkxtWAiL:1304691520) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/selector_impl.rs#L17)):*
I'm saying that, instead of:

pub type Stylist = ::style::selector_matching::Stylist<GeckoSelectorImpl>;

we do

pub type Stylist = style::selector_matching::Stylist<GeckoSelectorImpl>;

That compiles, right?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9567)*
<!-- Sent from Reviewable.io -->

emilio committed:
layout: Stop parameterizing on N for RecalcStyleAndConstructFlows

This is unfortunate, but making that useful would require parameterizing
`SharedLayoutContext` and `LayoutContext` depending on the
`SelectorImpl` (which is a **huge** work right now).

Probably the easier way to do it, and probably the one that keeps the
layout code more legible, and since there won't be multiple
implementations at the same compilation unit, would be "defining" a
default implementation for layout via feature flags.

That should allow us to remove the components/style/servo.rs file.

ecoal95 commented:
Review status: 27 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/layout/traversal.rs, line 32 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAGv97D4_Fb4h1eiPFX:-KAPcxRgM4ecrrMEKXd1:-1622810188) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/components/layout/traversal.rs#L32)):*
Humm... After giving it a quick try, there's *a lot* to change (since we should parameterize `LayoutContext`), including a lot of functions to parameterize due to that.

Thinking about it, most people shouldn't care about the `SelectorImpl` used in layout, and making layout reusable with a feature flag that defines the default `SelectorImpl` should be a lot more straightforward, given two style implementations won't coexist (which I don't think they should/will).

Anyways, that's not near-term, so I plan to land this, and work on querying anon-box pseudos from gecko. Removing the N parameter for now :P

---

*[ports/geckolib/selector_impl.rs, line 17 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9567#-KAHNDEnn4SaAOQNuUDe:-KAPerkWh2FQJGSsKclB:-1927556282) ([raw file](https://github.com/servo/servo/blob/76908598f843158aa49dd0daf5bb42e139b4b8f3/ports/geckolib/selector_impl.rs#L17)):*
Yes, given we add a `use style;` line at the top, that was what I was saying above.

Just did that :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9567)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9622) made this pull request unmergeable. Please resolve the merge conflicts.

bholley commented:
Looks good to me! Thanks for working on this. :-) r=bholley

---

Review status: 19 of 43 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9567#-:-KAROsnnzmgWaiSnq8Sd:-1430995830)*
<!-- Sent from Reviewable.io -->

bholley commented:
(specifically, on the layout and geckolib stuff. I looked at most of the style stuff, but not all of it, since @SimonSapin was already happy with it).

ecoal95 commented:
@bors-servo: r=bholley,SimonSapin

bors-servo commented:
:pushpin: Commit 04d2db5 has been approved by `bholley,SimonSapin`

<!-- @bors-servo r=bholley,SimonSapin 04d2db5747ce85b5ee7f9cbc9f6d13e523c68a76 -->

bors-servo commented:
:hourglass: Testing commit 04d2db5 with merge c11844c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1861)

ecoal95 commented:
@bors-servo: retry

* #8120 
* #8815 
* #7791
* #8191

(wow, so many intermittents hit this time)

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/1630), [gonk](http://build.servo.org/builders/gonk/builds/1663), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1636), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/336) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/1861), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1691), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1726)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1630), [gonk](http://build.servo.org/builders/gonk/builds/1663), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1636), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1862), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/336), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1691), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1727)
